File: tools/my_devices/climate.py
from __future__ import annotations

from tools.base import BaseTool, ToolResult
from tools.schemas.my_devices import AcControlInput


class AcControlTool(BaseTool):
    name = "ac_control"
    description = "에어컨을 제어합니다. 전원, 모드, 온도, 풍량을 설정할 수 있습니다."
    args_schema = AcControlInput

    def execute(self, args: AcControlInput) -> ToolResult:
        results: dict[str, object] = {}

        if args.power is not None:
            response = self.client.command(args.device_id, "switch", args.power)
            if isinstance(response, dict) and "error" in response:
                return ToolResult(
                    success=False,
                    error=response["error"].get("message", "Failed to set power"),
                    data=results,
                )
            results["power"] = response

        if args.mode is not None:
            response = self.client.command(
                args.device_id,
                "airConditionerMode",
                "setAirConditionerMode",
                [args.mode],
            )
            if isinstance(response, dict) and "error" in response:
                return ToolResult(
                    success=False,
                    error=response["error"].get("message", "Failed to set mode"),
                    data=results,
                )
            results["mode"] = response

        if args.temperature is not None:
            response = self.client.command(
                args.device_id,
                "thermostatCoolingSetpoint",
                "setCoolingSetpoint",
                [args.temperature],
            )
            if isinstance(response, dict) and "error" in response:
                return ToolResult(
                    success=False,
                    error=response["error"].get("message", "Failed to set temperature"),
                    data=results,
                )
            results["temperature"] = response

        if args.wind_level is not None:
            response = self.client.command(
                args.device_id,
                "airConditionerFanMode",
                "setFanMode",
                [str(args.wind_level)],
            )
            if isinstance(response, dict) and "error" in response:
                return ToolResult(
                    success=False,
                    error=response["error"].get("message", "Failed to set wind level"),
                    data=results,
                )
            results["wind_level"] = response

        return ToolResult(success=True, data=results)

File: smartthings_tools_rs/src/my_devices/climate.rs
use schemars::schema::RootSchema;
use schemars::schema_for;
use serde_json::Value;

use crate::client::SmartThingsClient;
use crate::common_tools::{extract_error_msg, has_error};
use crate::schemas::my_devices::AcControlInput;
use crate::tool::{Tool, ToolResult};

pub struct AcControlTool<'a> {
    pub client: &'a SmartThingsClient,
}

impl Tool for AcControlTool<'_> {
    fn name(&self) -> &'static str {
        "ac_control"
    }
    fn description(&self) -> &'static str {
        "에어컨을 제어합니다. 전원, 모드, 온도, 풍량을 설정할 수 있습니다."
    }
    fn parameters_schema(&self) -> RootSchema {
        schema_for!(AcControlInput)
    }

    fn execute(&self, args: Value) -> ToolResult {
        let input: AcControlInput = match serde_json::from_value(args) {
            Ok(v) => v,
            Err(e) => return ToolResult::fail(format!("ValidationError: {e}")),
        };

        let mut results = serde_json::Map::new();

        if let Some(ref power) = input.power {
            let data = self
                .client
                .command(&input.device_id, "switch", power, None, None);
            if has_error(&data) {
                return ToolResult::fail(extract_error_msg(&data));
            }
            results.insert(
                "power".to_string(),
                serde_json::to_value(&data).unwrap_or_default(),
            );
        }

        if let Some(ref mode) = input.mode {
            let data = self.client.command(
                &input.device_id,
                "airConditionerMode",
                "setAirConditionerMode",
                Some(vec![Value::String(mode.clone())]),
                None,
            );
            if has_error(&data) {
                return ToolResult::fail(extract_error_msg(&data));
            }
            results.insert(
                "mode".to_string(),
                serde_json::to_value(&data).unwrap_or_default(),
            );
        }

        if let Some(temp) = input.temperature {
            let data = self.client.command(
                &input.device_id,
                "thermostatCoolingSetpoint",
                "setCoolingSetpoint",
                Some(vec![serde_json::json!(temp)]),
                None,
            );
            if has_error(&data) {
                return ToolResult::fail(extract_error_msg(&data));
            }
            results.insert(
                "temperature".to_string(),
                serde_json::to_value(&data).unwrap_or_default(),
            );
        }

        if let Some(wind) = input.wind_level {
            let data = self.client.command(
                &input.device_id,
                "airConditionerFanMode",
                "setFanMode",
                Some(vec![Value::String(wind.to_string())]),
                None,
            );
            if has_error(&data) {
                return ToolResult::fail(extract_error_msg(&data));
            }
            results.insert(
                "wind_level".to_string(),
                serde_json::to_value(&data).unwrap_or_default(),
            );
        }

        ToolResult::ok(Value::Object(results))
    }
}
